x-pack/plugins/security_solution/server/lib/detection_engine/signals/utils.test.ts

```diff
@@ -609,134 +598,116 @@ describe('utils', () => {
     });
   });
 
-  describe('getSignalTimeTuples', () => {
+  describe('getRuleRangeTuples', () => {
     test('should return a single tuple if no gap', () => {
-      const someTuples = getSignalTimeTuples({
+      const { tuples, remainingGap } = getRuleRangeTuples({
         logger: mockLogger,
-        gap: null,
         previousStartedAt: moment().subtract(30, 's').toDate(),
         interval: '30s',
-        ruleParamsFrom: 'now-30s',
-        ruleParamsTo: 'now',
-        ruleParamsMaxSignals: 20,
+        from: 'now-30s',
+        to: 'now',
+        maxSignals: 20,
         buildRuleMessage,
       });
-      const someTuple = someTuples[0];
+      const someTuple = tuples[0];
       expect(moment(someTuple.to).diff(moment(someTuple.from), 's')).toEqual(30);
+      expect(tuples.length).toEqual(1);
+      expect(remainingGap.asMilliseconds()).toEqual(0);
+    });
+
+    test('should return a single tuple if malformed interval prevents gap calculation', () => {
+      const { tuples, remainingGap } = getRuleRangeTuples({
+        logger: mockLogger,
+        previousStartedAt: moment().subtract(30, 's').toDate(),
+        interval: 'invalid',
+        from: 'now-30s',
+        to: 'now',
+        maxSignals: 20,
+        buildRuleMessage,
+      });
+      const someTuple = tuples[0];
+      expect(moment(someTuple.to).diff(moment(someTuple.from), 's')).toEqual(30);
+      expect(tuples.length).toEqual(1);
+      expect(remainingGap.asMilliseconds()).toEqual(0);
     });
 
     test('should return two tuples if gap and previouslyStartedAt', () => {
-      const someTuples = getSignalTimeTuples({
+      const { tuples, remainingGap } = getRuleRangeTuples({
         logger: mockLogger,
-        gap: moment.duration(10, 's'),
         previousStartedAt: moment().subtract(65, 's').toDate(),
         interval: '50s',
-        ruleParamsFrom: 'now-55s',
-        ruleParamsTo: 'now',
-        ruleParamsMaxSignals: 20,
+        from: 'now-55s',
+        to: 'now',
+        maxSignals: 20,
         buildRuleMessage,
       });
-      const someTuple = someTuples[1];
-      expect(moment(someTuple.to).diff(moment(someTuple.from), 's')).toEqual(10);
+      const someTuple = tuples[1];
+      expect(moment(someTuple.to).diff(moment(someTuple.from), 's')).toEqual(55);
```

Comment:
So the tuples were reversed from the previous functionality, essentially?

Reply:
The previous implementation had a special case for if the gap was <1 interval which was the scenario in this test. In this special case it would create only one extra tuple which was scaled so it covered only the gap duration rather than a full rule interval duration. The refactored implementation always uses `interval + lookback` as the tuple duration for consistency, so the difference `to - from` here changed from 10s (the gap duration) to 55s (interval + lookback)

Reply:
I'm not sure this is the appropriate avenue to pursue. If a rule is scheduled to run weekly but only had a gap of an hour, it's going to schedule a second search over a weeks worth of data that it had already searched over instead of just the hours gap? That doesn't seem efficient.

Reply:
I left a [comment](https://github.com/elastic/kibana/pull/91966/files/3154687b9a88a7a3a5ac0b0c9e71dc6ba028308b#diff-333497786c31c4722a433be429554127c7eb40198062ef5004a057da5c0a23cfR501) in the code explaining some of the reasoning behind this change. The gist of it is that for some rule types a consistent query duration is important, and the overlap between consecutive rule runs affects the behavior as well. Since we have to be able to handle up to 4 extra gap-covering queries in a rule execution anyway I think the benefit of having consistent query durations outweighs the cost of extending the partial time duration to 1 full time duration.

Reply:
I would suggest re-working it in such a way that the default is the original functionality and the extended lookback is updated to use the overlap if the given rule type requires it. This way the lookback does not default to a full rule interval if the gap is less than that.

Reply:
Will the dupes be ignored? In the case of threshold, they will. If they're ignored for all rule types, it seems fine to me to leave the consistent query duration... if we're experiencing gaps frequently, then probably something else is wrong?

Reply:
All rule types have duplicate detection and must have it due to the expected overlap between query time ranges, so dupes will be ignored. @madirey my reasoning is the same, the gaps should be very infrequent so ensuring correctness by making the code easy to reason about was a higher priority to me than optimizing for each case.